tests/src/Integration/Action/UserRoleAddTest.php

```diff
@@ -0,0 +1,214 @@
+   */
+  public function testSummary() {
+    $this->assertEquals('Adds roles to a particular user', $this->action->summary());
+  }
+
+  /**
+   * Tests adding of one role to user. User should not be saved.
+   *
+   * @covers ::execute
+   */
+  public function testAddOneRoleNoSave() {
+    // Set-up a mock user.
+    $account = $this->getMockedUser();
+    $account->expects($this->once())
+      ->method('hasRole')
+      ->with($this->equalTo('administrator'))
+      ->will($this->returnValue(FALSE));
+    $account->expects($this->once())
+      ->method('addRole')
+      ->with($this->equalTo('administrator'));
+    // We do noe expect call of the 'save' method because user should be
+    // auto-saved later.
+    $account->expects($this->never())
+      ->method('save');
+
+    // Mock the 'administrator' user role.
+    $administrator = $this->getMockedUserRole('administrator');
+
+    // Test adding of one role.
+    $this->action
+      ->setContextValue('user', $account)
+      ->setContextValue('roles', [$administrator])
+      ->execute();
+  }
+
+  /**
+   * Tests adding of three roles to user.
+   *
+   * @covers ::execute
+   */
+  public function testAddThreeRoles() {
+    // Set-up a mock user.
+    $account = $this->getMockedUser();
+    // Mock hasRole.
+    $account->expects($this->exactly(3))
+      ->method('hasRole')
+      ->with($this->logicalOr(
+        $this->equalTo('manager'),
+        $this->equalTo('editor'),
+        $this->equalTo('administrator')
+      ))
+      ->will($this->returnValue(FALSE));
+    // Mock addRole.
+    $account->expects($this->exactly(3))
+      ->method('addRole')
+      ->with($this->logicalOr(
+        $this->equalTo('manager'),
+        $this->equalTo('editor'),
+        $this->equalTo('administrator')
+      ));
+
+    // Mock user roles.
+    $manager = $this->getMockedUserRole('manager');
+    $editor = $this->getMockedUserRole('editor');
+    $administrator = $this->getMockedUserRole('administrator');
+
+    // Test adding of three roles role.
+    $this->action
+      ->setContextValue('user', $account)
+      ->setContextValue('roles', [$manager, $editor, $administrator])
+      ->execute();
+  }
+
+  /**
+   * Tests adding of existing role to user.
+   *
+   * @covers ::execute
+   */
+  public function testAddExistingRole() {
+    // Set-up a mock user with role 'administrator'.
+    $account = $this->getMockedUser();
+    $account->expects($this->once())
+      ->method('hasRole')
+      ->with($this->equalTo('administrator'))
+      ->will($this->returnValue(TRUE));
+
+    // We do not expect a call of the 'addRole' method.
+    $account->expects($this->never())
+      ->method('addRole');
+
+    // Mock the 'administrator' user role.
+    $administrator = $this->getMockedUserRole('administrator');
+
+    // Test adding one role.
+    $this->action
+      ->setContextValue('user', $account)
+      ->setContextValue('roles', [$administrator])
+      ->execute();
```

Comment:
After that we need to call autoSaveContext() on the plugin and make sure that it does NOT return ['user'].